### PR TITLE
Ability to setup aliases for entities spawned from spawnables.

### PR DESCRIPTION
**Pull request by AMZN-koppersr**
Added support for setting up aliases for entities spawned from spawnables through the Spawnable Entities Interface/Manager. 
Updates in preparation for adding entity aliases to spawnables. Entity aliases can be used to have a request to spawn an entity:
- spawn the original entity as normal
- be disabled
- redirected to another entity in another spawnable
- also spawn an entity from another spawnable
- add the components from an entity in another spawnable

An entity alias can indicate whether or not to load the spawnable dependency. If the spawnable dependency is loaded it will be loaded asynchronously because starting blocking loads in an asset handler can lead to deadlocks once there are no more jobs available to deserialize assets.

Entity aliases can be updated as a reaction to the spawnable being loaded or at any other time afterwards through the Spawnable Entities Interface. Currently these changes are applied to the spawnable that owns the entity aliases, but once the Spawnable Entities Interface makes use of AzFramework::Scene a copy of the entity aliases should be stored in the scene and be updated instead of the spawnable.

The Prefab Processor Stack has been updated to support setting up entity aliases and resolve spawnable assets at a later time.

Unit tests were added to cover the loading and spawning of entities from spawnables that use aliases.

The following additional changes were made:
- AssetDataStream can now return the stored streaming deadline and priority.
- RootSpawnable now has an event that's called just before root spawnable spawns entities. This is an immediate event unlike the other events that are queued.
- The return values in from the processing functions in the Spawnable Entities Manager now have a clearer return value to indicate whether a request has completed or is being re-queued.
- A load barrier was added to the Spawnable Entities Interface, which acts the same as a regular barrier but also accounts for the spawnable being loaded and won't trigger the callback until has completed.

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableAssetHandler.cpp [hunk 1]
```diff
@@ -91,4 +94,40 @@ namespace AzFramework
...
+                    [](const Spawnable::EntityAlias& lhs, const Spawnable::EntityAlias& rhs)
+                    {
+                        return lhs.HasLowerIndex(rhs);
+                    }),
+                "Spawnable '%s' has an unsorted entity alias list.", asset.GetHint().c_str());
+
+            SpawnableAssetEventsBus::Broadcast(
+                &SpawnableAssetEvents::OnResolveAliases, aliases, spawnable->GetMetaData(), spawnable->GetEntities());
+
+            aliases.Optimize();
```
Comment (NicholasVanSickle) [inline Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableAssetHandler.cpp:121]: Might be worth a comment just to point out that this is a no-op if the alias list isn't dirty after `OnResolveAliases`

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesContainer.h [hunk 2]
```diff
@@ -73,7 +79,11 @@ namespace AzFramework
         //! other than the calling thread including the main thread. Note that because the alert is queued it can still be called
         //! after the container has been deleted or can be called for a previously assigned spawnable. In the latter case check
         //! if the current generation matches the generation provided with the callback.
-        void Alert(AlertCallback callback);
+        //! @callback The function called when the alert triggers. This can be called from a different thread than the one that
+        //!     the one that made the call to Alert.
+        //! @checkSpawnableIsLoaded If true the alert will also block until the spawnable has been loaded. If false then it will
```
Comment (NicholasVanSickle) [inline Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesContainer.h:84]: nit: `@param callback` & `@param checkSpawnableIsLoaded`

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h [hunk 3]
```diff
@@ -29,7 +30,143 @@ namespace AzFramework
...
+            ~EntityAliasVisitor();
+
+            EntityAliasVisitor(EntityAliasVisitor&& rhs);
+            EntityAliasVisitor& operator=(EntityAliasVisitor&& rhs);
+
+            EntityAliasVisitor(const EntityAliasVisitor& rhs) = delete;
+            EntityAliasVisitor& operator=(const EntityAliasVisitor& rhs) = delete;
+
+            //! Checks if the visitor was able to retrieve data. This needs to be checked before calling any other functions.
+            bool IsSet() const;
```
Comment (NicholasVanSickle) [inline Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:103]: This might read better as `CanReadAliases`? In practice this seems to fail when we fail to acquire the entiy alias list in the `Try` methods so I think it might read easier if that's directly implied by checking this.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h [hunk 4]
```diff
@@ -29,7 +30,143 @@ namespace AzFramework
...
+
+            bool HasAliases() const;
+            bool AreAllSpawnablesReady() const;
+
+            EntityAliasList::const_iterator begin() const;
```
Comment (NicholasVanSickle) [inline Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:108]: A guide comment for "the list itself is immutable during iteration so these are const iterators" would prevent some light astonishment here

**Review by NicholasVanSickle — COMMENTED** [review]
Lot to get through here, but this looks really good! I did a pass on the alias authoring side of this around `EntityAliasVisitor` et al, I'll follow up on Monday to get this fully reviewed

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h [hunk 5]
```diff
@@ -29,7 +30,143 @@ namespace AzFramework
...
         AZ_RTTI(AzFramework::Spawnable, "{855E3021-D305-4845-B284-20C3F7FDF16B}", AZ::Data::AssetData);
 
+        // The order is important for sorting in the SpawnableAssetHandler.
+        enum class EntityAliasType : uint8_t
+        {
+            Original,   //!< The original entity is spawned.
+            Disabled,   //!< No entity will be spawned.
+            Replace,    //!< The entity alias is spawned instead of the original.
+            Additional, //!< The original entity is spawned as well as the alias. The alias will get a new entity id.
+            Merge       //!< The original entity is spawned and the components of the alias are added. The caller is responsible for
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:40]: As these are actions to take "Replace" and "Merge" are more appropriate. "Disabled" has been updated to "Disable" to be in line with this.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.cpp [hunk 6]
```diff
@@ -6,12 +6,479 @@
...
+        rhs.m_entityAliasList = nullptr;
+        rhs.m_dirty = false;
+    }
+
+    auto Spawnable::EntityAliasVisitor::operator=(EntityAliasVisitor&& rhs) -> EntityAliasVisitor&
+    {
+        if (this != &rhs)
+        {
+            this->~EntityAliasVisitor();
+            *this = EntityAliasVisitor(rhs.m_owner, rhs.m_entityAliasList);
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.cpp:170]: I was trying to avoid placement new because that seems to confuse a lot of people, but the alternative doesn't seem to be much better in that regard. Going to replace this with a placement new as that will also avoid some duplicate code.

File: Code/Framework/AzFramework/AzFramework/Spawnable/RootSpawnableInterface.h [hunk 7]
```diff
@@ -30,12 +30,22 @@ namespace AzFramework
...
+        //! Called when the root spawnable has completed spawning of entities. This may be called several times without a call to release
+        //!     in between.
+        //! NOTE: This callback is queued and will be called with a delay and from the main thread.
+        //! @param rootSpawnable The new root spawnable that was used to spawn entities from.
+        //! @param generation The generation of the root spawnable. This will increment every time a new spawnable is assigned.
+        virtual void OnRootSpawnableReady(
+            [[maybe_unused]] AZ::Data::Asset<Spawnable> rootSpawnable, [[maybe_unused]] uint32_t generation) {}
         //! Called when the root spawnable has Released. This will only be called if there's no root spawnable assigned to take the
         //!     place of the original root spawnable.
+        //! Note: This callback is queued and will be called with a delay and from the main thread.
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/RootSpawnableInterface.h:48]: Neat, will do.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h (hunk 3 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:103]: I'm not a fan of `IsSet` either, however your suggestion also doesn't do it for me. The problem is that there are two iterators which would end up in `CanReadAliases` and `CanReadWriteAliases` and I would prefer having a single function. Also I think this can be extended to update more than just aliases, like the meta data, in the future. I didn't want to add even more to this PR as it's already pretty big but that would also mean that the "Aliases" part would have to change in the future.
In other words, I'm not sure what a good name for this function would be, so any suggestions are welcome.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.cpp [hunk 8]
```diff
@@ -6,12 +6,479 @@
...
+#include <AzCore/std/typetraits/typetraits.h>
 #include <AzFramework/Spawnable/Spawnable.h>
 
 namespace AzFramework
 {
+    //
+    // EntityAlias
+    //
+
+
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.cpp:23]: Ok, although above those comments I do prefer to keep 2 or 3 blank lines for readability.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h [hunk 9]
```diff
@@ -29,7 +30,143 @@ namespace AzFramework
...
+                        //!< maintaining a valid component list.
+        };
+
+        enum ShareState : int32_t
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:44]: Yes, negative values indicate the spawnable is read-only shared, which can be for multiple visitors at the same time. As such this can become larger that -127.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h [hunk 10]
```diff
@@ -29,7 +30,143 @@ namespace AzFramework
...
+        {
+            Read = -1,
+            NotShared = 0,
+            ReadWrite = 1
+        };
+
+        //! An entity alias redirects the spawning of an entity to another entity, possibly in another spawnable.
+        struct EntityAlias
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:52]: To use this it requires typing `Spawnable::EntityAlias` because it's a nested struct, so adding `Spawnable` would just duplicate the same information.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableAssetHandler.cpp [hunk 11]
```diff
@@ -91,4 +94,40 @@ namespace AzFramework
         AZ::Uuid subIdHash = AZ::Uuid::CreateData(id.data(), id.size());
         return azlossy_caster(subIdHash.GetHash());
     }
+
+    void SpawnableAssetHandler::ResolveEntityAliases(
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableAssetHandler.cpp:98]: It does more than just optimize. It allows any external handlers to make updates to the aliases (which may require doing another optimization) and then start loading of spawnables that were marked to do so.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 12]
```diff
@@ -316,161 +403,264 @@ namespace AzFramework
...
+                                    previousEntity = clone;
+                                    if (clone)
+                                    {
+                                        spawnedEntities.emplace_back(clone);
+                                        spawnedEntityIndices.push_back(index);
+                                    }
+
+                                    ++aliasIt;
+                                } while (aliasIt != aliasEnd && aliasIt->m_sourceIndex == index);
+                            }
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:621]: can you move this into a helper function because its the same code as in SpawnAllEntities() too.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 13]
```diff
@@ -495,15 +685,15 @@ namespace AzFramework
             }
 
             ticket.m_currentRequestId++;
-            return true;
+            return CommandResult::Executed;
         }
         else
         {
-            return false;
+            return CommandResult::Requeue;
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:692]: Thanks for this. This makes it obvious that the request will be requeued.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 14]
```diff
@@ -690,35 +905,64 @@ namespace AzFramework
...
 
-    bool SpawnableEntitiesManager::ProcessRequest(DestroyTicketCommand& request)
+    auto SpawnableEntitiesManager::ProcessRequest(DestroyTicketCommand& request) -> CommandResult
     {
         if (request.m_requestId == request.m_ticket->m_currentRequestId)
         {
             for (AZ::Entity* entity : request.m_ticket->m_spawnedEntities)
             {
-                if (entity != nullptr)
+                if (entity != nullptr && !entity->GetComponents().empty())
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:946]: Can you explain why you added this? Because an entity can exist without any components. Particularly if users write custom scripts/code to remove all components from entity after spawning. I can see that you are deleting the entity below if components are empty. But deleting the entity is not the same as destroying in the game context since that will send other entity destruction related ebus calls too

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp [hunk 15]
```diff
@@ -228,6 +231,23 @@ namespace AzToolsFramework
             m_entities.clear();
         }
 
+        AZStd::unique_ptr<AZ::Entity> Instance::ReplaceEntity(AZStd::unique_ptr<AZ::Entity>&& entity, EntityAliasView alias)
+        {
+            AZStd::unique_ptr<AZ::Entity> result;
+            auto it = m_entities.find(alias);
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp:237]: Don't you need to look for the alias in the container entity too?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp [hunk 16]
```diff
@@ -377,7 +397,12 @@ namespace AzToolsFramework
             return entityAliases;
         }
 
-        void Instance::GetNestedEntityIds(const AZStd::function<bool(AZ::EntityId)>& callback)
+        size_t Instance::GetEntityAliasCount() const
+        {
+            return m_entities.size();
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp:402]: same. Don't you need to account for container entity too? So this should be size +1

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (srikappa-amzn) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:52]: That's true. However, I can see some confusion arising when discussing entity aliases for prefabs and spawnables. And I can already see ourselves calling them PrefabEntityAlias and SpawnableEntityAlias to resolve ambiguity during discussions. Hence, my suggestion to name it SpawnableEntityAlias now itself because changing it later will prove challenging since the word 'entityAlias' will be used in many files at that point and we need to be careful where we change the name then.

File: Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp [hunk 17]
```diff
@@ -754,6 +1011,190 @@ namespace UnitTest
...
+                {
+                    allReplaced = allReplaced && entity->FindComponent<SourceSpawnableComponent>() == nullptr;
+                    allReplaced = allReplaced && entity->FindComponent<TargetSpawnableComponent>() != nullptr;
+                }
+                else
+                {
+                    allReplaced = false;
+                }
+            }
+        };
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp:1095]: can this be moved to a helper function since its repeated?

File: Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp [hunk 18]
```diff
@@ -754,6 +1011,190 @@ namespace UnitTest
...
+                        allReplaced = allReplaced && entity->FindComponent<TargetSpawnableComponent>() != nullptr;
+                    }
+                    onSource = !onSource;
+                }
+                else
+                {
+                    allReplaced = false;
+                }
+            }
+        };
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp:1147]: can this be moved to a helper function since its repeated?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Spawnable/SpawnableUtils.cpp [hunk 19]
```diff
@@ -47,6 +165,122 @@ namespace AzToolsFramework::Prefab::SpawnableUtils
...
+            store.m_target.emplace<EntityAliasSpawnableLink>(target, result->GetId());
+            store.m_tag = tag;
+            store.m_loadBehavior = loadBehavior;
+            context.RegisterSpawnableEntityAlias(AZStd::move(store));
+
+            return result;
+        }
+        else
+        {
+            AZ_Assert(false, "Entity with id %zu was not found in the source prefab.", static_cast<AZ::u64>(entity));
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Spawnable/SpawnableUtils.cpp:239]: nit: the cast should be to `size_t` here to match the print specifier (or use `PRIx64` in the format string)

**Review by NicholasVanSickle — APPROVED** [review]
I've at least given everything a once-over; a few nits (and I generally agree with the other feedback) but I think this is a good starting point as a general purpose hook layer for mutating spawnable behavior so I'm giving this an approve.

It seems like in many cases aliases can be used to spawn a root spawnable instead of a new spawnable type (like we have for network spawnables currently). It might be helpful to draft some best practices around when the spawnable alias system should and should not be used so that can give sensible recommendations.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h (hunk 3 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:103]: Went with `IsValid`.

File: Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.h:52]: I don't think turning `Spawnable::EntityAlias` into `Spawnable::SpawnableEntityAlias` will change it being referenced as "entity aliases" in discussions rather than "Spawnable entity aliases" and "Prefab entity aliases". There's also not a similar prefix for spawnable entities and prefab entities, which are more prone to confusion imho.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 20]
```diff
@@ -280,7 +307,67 @@ namespace AzFramework
         constexpr bool allowDuplicateIds = false;
 
         return AZ::IdUtils::Remapper<AZ::EntityId, allowDuplicateIds>::CloneObjectAndGenerateNewIdsAndFixRefs(
-                &entityTemplate, templateToCloneMap, &serializeContext);
+            &entityTemplate, templateToCloneMap, &serializeContext);
+    }
+
+    AZ::Entity* SpawnableEntitiesManager::CloneSingleAliasedEntity(
+        const AZ::Entity& entityTemplate,
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:314]: I'll use "prototype" instead then, so `entityPrototype` as the entity passed in is used as template/prototype to create the actual entity with.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 21]
```diff
@@ -280,7 +307,67 @@ namespace AzFramework
...
+        AZ::SerializeContext& serializeContext)
+    {
+        using ResultType = AZStd::pair<AZ::Entity*, uint32_t>;
+
+        AZ::Entity* clone = nullptr;
+        switch (alias.m_aliasType)
+        {
+        case Spawnable::EntityAliasType::Original:
+            // Behave as the original version.
+            clone = CloneSingleEntity(entityTemplate, templateToCloneMap, serializeContext);
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:327]: If the alias type is original than `alias.m_spawnable` is null. An asset cannot point to itself because it would cause a memory leak. When an asset refers to itself it will keep a reference count to itself at all times so the asset never gets cleared.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 22]
```diff
@@ -316,161 +403,264 @@ namespace AzFramework
...
+                    for (uint32_t i = 0; i < entitiesToSpawnSize; ++i)
+                    {
+                        // If this entity has previously been spawned, give it a new id in the reference map
+                        RefreshEntityIdMapping(
+                            entitiesToSpawn[i].get()->GetId(), ticket.m_entityIdReferenceMap, ticket.m_previouslySpawned);
 
-                spawnedEntities.emplace_back(clone);
-                spawnedEntityIndices.push_back(i);
-            }
+                        if (aliasIt == aliasEnd || aliasIt->m_sourceIndex != i)
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:459]: The first check is to see if there are no aliases at all, in which case a cheaper code path is used that doesn't consider aliases. The second check is part of a loop that iterates over the aliases so `aliasIt` keeps moving forward which can result in it reaching the end if one or more entities at the end of source spawnable doesn't have an alias.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp [hunk 23]
```diff
@@ -316,161 +403,264 @@ namespace AzFramework
...
+                                    *entitiesToSpawn[i], *aliasIt, ticket.m_entityIdReferenceMap, previousEntity,
+                                    *request.m_serializeContext);
+                                previousEntity = clone;
+                                if (clone)
+                                {
+                                    spawnedEntities.emplace_back(clone);
+                                    spawnedEntityIndices.push_back(i);
+                                }
+                                ++aliasIt;
+                            } while (aliasIt != aliasEnd && aliasIt->m_sourceIndex == i);
```
[the remark this answers is not in the document]
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:482]: There are two types of aliases, replacement (original, disabled, replace) and additions (merge, additional). The optimization phase already removes all duplicate aliases so there can only ever be one replacement entity. There can however be multiple additions as these either add components to the entity that was spawned as the replacement or spawns an additional entity.

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp (hunk 12 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:621]: The shared code has already been factored into functions. The remaining code has some subtle differences because `SpawnAllEntities` can use a straight loop while `SpawnEntities` uses a random list of indices. I did try to split this up into further functions but it didn't help readability. To make it readable would come at the cost of performance which for entity spawning isn't worth taking DRY to an extreme imho.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp (hunk 15 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp:237]: The entities are flattened for spawnables so there's no explicit container entity.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp (hunk 16 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp:402]: Same

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp (hunk 14 above)
Comment (srikappa-amzn) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:947]: Looks like you missed answering this. Commenting on this again to get visibility

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp (hunk 15 above)
Comment (srikappa-amzn) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.cpp:237]: There maybe no explicit container entity for the spawnable. But a user provides the alias that corresponds to the container entity of the prefab instance, then it won't be available under 'm_entities'. So, instead of replacing the container entity and returning the original one, you will be returning a nullptr.

File: Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp [hunk 24]
```diff
@@ -366,33 +542,144 @@ namespace UnitTest
...
+        FillSpawnable(NumEntities);
+        AZ::Data::Asset<Spawnable> target = CreateTargetSpawnable(4);
+        InsertEntityAliases<4>(
+            { 0, 1, 2, 3 }, { 0, 1, 2, 3 },
+            { Spawnable::EntityAliasType::Merge, Spawnable::EntityAliasType::Merge, Spawnable::EntityAliasType::Merge,
+              Spawnable::EntityAliasType::Merge },
+            &target);
+
+        size_t spawnedEntitiesCount = 0;
+        bool allMerged = true;
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/Tests/Spawnable/SpawnableEntitiesManagerTests.cpp:668]: repeated : recommend changing the default to 'false'. That way, later when you do 'EXPECT_TRUE(allMerged)', we can be sure that the lambda actually worked and flipped the bool to true. Otherwise, if the call to SEM fails, the bool can result in a false positive.

File: Code/Framework/AzFramework/Tests/Spawnable/SpawnableTests.cpp [hunk 25]
```diff
@@ -0,0 +1,513 @@
...
+
+    //
+    // UpdateAliasType
+    //
+
+    TEST_F(SpawnableTest, EntityAliasVisitor_UpdateAliasType_AllToOriginal_NoAliasesAfterOptimization)
+    {
+        using namespace AzFramework;
+        InsertEntities(8);
+        InsertEntityAliases<8>(
```
Comment (srikappa-amzn) [inline Code/Framework/AzFramework/Tests/Spawnable/SpawnableTests.cpp:302]: repeated nit: Can these be replaced with 'DefaultEntityAliasTestCount' as well?

File: Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp (hunk 14 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp:947]: The check was added because entities are not spawned it doesn't have any components. This is because several components depend on entities having at least a transform component. Because those entities are not spawned there's also no reason to destroy them here because it would be referring to non-existing entity (at least from the game context point of view).

File: Code/Framework/AzFramework/Tests/Spawnable/SpawnableTests.cpp (hunk 25 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzFramework/Tests/Spawnable/SpawnableTests.cpp:302]: I'd rather not because those each need to provide arguments with exactly that number of entries. Having the exact number there makes it more readable imho than have `DefaultEntityAliasTestCount`.

**Review by srikappa-amzn — APPROVED** [review]
